### PR TITLE
BLD: Build wheels on cibuildwheel

### LINKED ISSUE #3916: [ENH] CPython 3.9 Windows and macOS wheels
Would it be possible to have Python 3.9 wheels for Windows and macOS too?

Related to #3868:

> **Is your feature request related to a problem? Please describe.**
> A clear and concise description of what the problem is. Ex. In my code, I would like to do [...]
> 
> ```
> $ python3.9 -m pip install cython
> ```
> 
> I'd like this to be super fast.
> 
> **Describe the solution you'd like**
> A clear and concise description of what you want to happen, including code examples if applicable.
> 
> I'd like pip to install Cython wheels with Python 3.9.
> 
> **Describe alternatives you've considered**
> 
> It is possible to install Cython with `--install-option="--no-cython-compile"`, but that doesn't work in situations where Cython is a dependency.

Thanks

### LINKED ISSUE #4736: [ENH] Switch to cibuildwheel
**Is your feature request related to a problem? Please describe.**
The wheel building process in Cython is currently clunky and is missing/not building wheels(other than the universal wheel, which I don't think is compiled) for any platforms other than Linux. Migrating to cibuildwheel will provide better release automation and build isolation and make building wheels easier(it's pretty easy to add new platforms using cibuildwheel as long as tests pass). 

It also saves end users time, as although Cython still works, they still need to compile it upon install (which takes a fair bit of time).

**Describe the solution you'd like**

I would like to see Cython migrate to cibuildwheel for release 3.1(but the infra for cibuildwheel can be put in much earlier than that). Unfortunately, cibuildwheel seems to be pretty stringent about dropping Python versions, and they tend to drop them on the dot when they become EOL. This may be a problem for Cython, if they wish to support any Python minor version past EOL. Pinning cibuildwheel to an old version doesn't work, as cibuildwheel as the list of platforms/Python versions it supports is baked into a particular version, meaning that Cython will not be able to support any new platforms/Python versions that cibuildwheel adds support for if they pin.

My recommendation to solve this is to first do the work adding the infrastructure to build in cibuildwheel, but leave it dormant and continue using the current system to build wheels. After 3.0, which deprecates Python 2.7, there probably should be little reason to support EOL python versions. When Cython 3.1 releases, which drops 2.7 support, we can then activate the cibuildwheel infra.

**Describe alternatives you've considered**
multibuild, but cibuildwheel is better as it has better support for building wheels for pre-release Pythons, and doesn't need to be vendored as a submodule.

**Additional context**

I am willing to put up a PR for this if this is welcome.

cc @scoder @da-woods

### PR DESCRIPTION
**Pull request by lithomas1**
closes #4736
closes #3916
Initial commits to build wheels and sdist on cibuildwheel.
Only Python > 3.6 is supported with this(hopefully that is also the case for Cython 3.1).

### CODE REVIEW

File: .github/workflows/wheels.yml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        with:
+          submodules: true
+          # versioneer.py requires the latest tag to be reachable. Here we
+          # fetch the complete history to get access to the tags.
+          # A shallow clone can work when the following issue is resolved:
+          # https://github.com/actions/checkout/issues/338
+          fetch-depth: 0
+      
+      - name: Build wheels
+        uses: pypa/cibuildwheel@v2.4.0
```
Comment (EwoutH) [inline .github/workflows/wheels.yml:56]: v2.5.0 is the most recent tag
```suggestion
        uses: pypa/cibuildwheel@v2.5.0
```

File: .github/workflows/wheels.yml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    name: Build wheel for ${{ matrix.python }}-${{ matrix.buildplat[1] }}
+    runs-on: ${{ matrix.buildplat[0] }}
+    strategy:
+      # Ensure that a wheel builder finishes even if another fails
+      fail-fast: false
+      matrix:
+        # Github Actions doesn't support pairing matrix values together, let's improvise
+        # https://github.com/github/feedback/discussions/7835#discussioncomment-1769026
+        buildplat:
+        - [ubuntu-20.04, manylinux_x86_64]
```
Comment (EwoutH) [inline .github/workflows/wheels.yml:38]: Let's use the most recent environments, using `-latest` (`ubuntu-latest`, etc.)

File: .github/workflows/wheels.yml [hunk 3]
```diff
@@ -0,0 +1,97 @@
+# Workflow to build wheels for upload to PyPI.
+# While this workflow is not active right now,
+# it will be used for Cython 3.1 after Python 2.7
+# support is dropped.
+name: Wheel Builder
+on:
+  schedule:
```
Comment (EwoutH) [inline .github/workflows/wheels.yml:7]: Do we really want to build wheels every push and PR to master?

File: .github/workflows/wheels.yml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+          # https://github.com/actions/checkout/issues/338
+          fetch-depth: 0
+      # Used to push the built wheels
+      - uses: actions/setup-python@v3
+        with:
+          # Build sdist on lowest supported Python
+          python-version: '3.6'
+      - name: Build sdist
+        run: |
+          python setup.py sdist
```
Comment (EwoutH) [inline .github/workflows/wheels.yml:93]: Calling setup.py directly is depreciated, I would switch to using pypa's [build](https://github.com/pypa/build) tool for wheel and sdist building.

See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for reasoning and https://pypa-build.readthedocs.io/en/stable/ for build documentation.

```suggestion
      - run: pip install --upgrade build wheel setuptools
      - run: python -m build --sdist
```

**Review by EwoutH — COMMENTED** [review]
Thanks for this PR, I was also thinking that it would be nice to switch to cibuildwheel for wheel builds. I did leave a few comments here and there.

Did you copy this template from somewhere by the way?

File: .github/workflows/wheels.yml (hunk 3 above)
Comment (scoder) [reply .github/workflows/wheels.yml:7]: Currently, wheel building is a release-only workflow. I think we can keep it that way.

File: .github/workflows/wheels.yml (hunk 2 above)
Comment (scoder) [reply .github/workflows/wheels.yml:38]: Should be ok, yes. The Linux wheel builds all use Docker, so C compilers etc. aren't an issue for this. Latest should be fine.

File: .github/workflows/wheels.yml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+          submodules: true
+          # versioneer.py requires the latest tag to be reachable. Here we
+          # fetch the complete history to get access to the tags.
+          # A shallow clone can work when the following issue is resolved:
+          # https://github.com/actions/checkout/issues/338
+          fetch-depth: 0
+      # Used to push the built wheels
+      - uses: actions/setup-python@v3
+        with:
+          # Build sdist on lowest supported Python
```
Comment (scoder) [inline .github/workflows/wheels.yml:89]: That would be 2.7, but I think we built our sdists on some not-so-old Python 3.x for quite a while now.

File: .github/workflows/wheels.yml (hunk 4 above)
Comment (lithomas1) [reply .github/workflows/wheels.yml:93]: Hm. Cython doesn't have a pyproject.toml file yet. I would wait until we have that available. 
It looks like build will just execute setup.py under the hood, which is equivalent to what we're doing now without it.
I'd like to remain explicit about the setup.py usage until we get a pyproject.toml.

**Comment by lithomas1** [comment]
> Thanks for this PR, I was also thinking that it would be nice to switch to cibuildwheel for wheel builds. I did leave a few comments here and there.
> 
> Did you copy this template from somewhere by the way?

It's from work I've been doing in the numpy repo but stripped down a lot. I'll add a link back in my next commit.

File: .github/workflows/wheels.yml (hunk 5 above)
Comment (lithomas1) [reply .github/workflows/wheels.yml:89]: Yeah, I was thinking in terms of post 3.0 when we wouldn't have to worry about Python 2.7. Are you planning to use the sdist part of this for 3.0? I can update this in that case.

File: .github/workflows/wheels.yml (hunk 3 above)
Comment (lithomas1) [reply .github/workflows/wheels.yml:7]: I would advocate for a weekly build in addition to working on release. From my work on the wheels for numpy/pandas, these things tend to break from time to time. 

In numpy in particular, we also can trigger builds using a label or a certain commit message, to run the wheel builder CI when a PR changes the build config. If Cython wants this, I can also port this over too.

Comment (scoder) [reply .github/workflows/wheels.yml:7]: Sounds great. Let's have weekly builds. We can even archive the wheels, as long as the release upload keeps being limited to releases.

Also +1 for supporting at least a label as trigger. Being able to add it to build related PRs would be really handy.

Comment (da-woods) [reply .github/workflows/wheels.yml:7]: We've occasionally had hard to diagnose problems with the shared module and changes to `CyFunction` between releases. We might see that more with regular pre-release builds, especially if they're used by a big project like Numpy.

It's probably solvable though (we could add the hashes of the utility code files to the shared API name).

Comment (lithomas1) [reply .github/workflows/wheels.yml:16]: > Sounds great. Let's have weekly builds. We can even archive the wheels, as long as the release upload keeps being limited to releases.

The scientific python packages have been uploading their nightlies to https://anaconda.org/scipy-wheels-nightly/repo. We can probably upload our wheels there too, if you want. I think @mattip has the API token for this.

> 
> Also +1 for supporting at least a label as trigger. Being able to add it to build related PRs would be really handy.

I've currently set the label to "Build System". LMK if you want to use another label.

**Comment by lithomas1** [comment]
This should be reader for another review now. Can someone please apply the "Build System" label to this PR to run the wheel builders?

**Comment by da-woods** [comment]
> lithomas1 requested review from scoder and da-woods 20 hours ago

I don't think I have any hugely useful comments. There's a slight conflict between:
* we want people to test Cython changes regularly and early
* Too many people using non-release versions probably has a support cost (as in my comment about the shared ABI).
It's probably worthwhile though. It probably should also end up being applied to the 0.29.x branch

**Comment by lithomas1** [comment]
gentle ping @scoder.

**Comment by scoder** [comment]
What about the aarch64 and i686 wheel builds?

File: .github/workflows/wheels.yml [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    name: Build sdist
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Cython
+        uses: actions/checkout@v3
+      # Used to push the built wheels
+      - uses: actions/setup-python@v3
+        with:
+          # Build sdist on lowest supported Python
+          python-version: '3.6'
```
Comment (scoder) [inline .github/workflows/wheels.yml:86]: Let's use something that is at least a little more recent and not out of maintenance yet.
```suggestion
          python-version: '3.8'
```

**Comment by lithomas1** [comment]
> What about the aarch64 and i686 wheel builds?

Are you still in the negative in Travis CI credits? I've added i686 wheel builds, but GHA doesn't have a native aarch64 platform, so we'd have to consider different CI platforms or emulating it with QEMU.

**Comment by scoder** [comment]
> consider emulating [aarch64] with QEMU.

That's what we currently do. Looks like cibuildwheel supports it as well:
https://cibuildwheel.readthedocs.io/en/stable/faq/#emulation

File: .github/workflows/wheels.yml [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        - [macos-10.15, macosx_*]
+        - [windows-2019, win_amd64]
+        - [windows-2019, win32]
+        python: ["cp36", "cp37", "cp38", "cp39", "cp310"] # Note: Wheels not needed for PyPy
+    steps:
+      - name: Checkout Cython
+        uses: actions/checkout@v3
+      
+      - name: Build wheels
+        uses: pypa/cibuildwheel@v2.5.0
```
Comment (scoder) [inline .github/workflows/wheels.yml:60]: We still need to set the `CIBW_ARCHS_LINUX` env var based on the target architecture, according to https://cibuildwheel.readthedocs.io/en/stable/faq/#emulation
We could probably also avoid the QEMU start on i686 and x86_64.
```suggestion
      - name: Checkout Cython
        uses: actions/checkout@v3

      - name: Set up QEMU
        if: runner.os == 'Linux'
        uses: docker/setup-qemu-action@v1
        with:
          platforms: all

      - name: Build wheels
        uses: pypa/cibuildwheel@v2.5.0
```

File: .github/workflows/wheels.yml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      matrix:
+        # Github Actions doesn't support pairing matrix values together, let's improvise
+        # https://github.com/github/feedback/discussions/7835#discussioncomment-1769026
+        buildplat:
+        - [ubuntu-20.04, manylinux_x86_64]
+        - [ubuntu-20.04, musllinux_x86_64]
+        - [macos-10.15, macosx_*]
+        - [windows-2019, win_amd64]
+        - [windows-2019, win32]
+        python: ["cp36", "cp37", "cp38", "cp39", "cp310"] # Note: Wheels not needed for PyPy
```
Comment (scoder) [inline .github/workflows/wheels.yml:54]: > Wheels not needed for PyPy

This reminds me that we also need a pure Python (2/3) wheel for PyPy (and other platforms for which we don't build binary wheels).

Comment (lithomas1) [reply .github/workflows/wheels.yml:55]: Unfortunately, cibuildwheel doesn't support this and doesn't intend to in the future.

I can refactor the sdist job to also build a pure python wheel.

**Comment by lithomas1** [comment]
> > consider emulating [aarch64] with QEMU.
> 
> That's what we currently do. Looks like cibuildwheel supports it as well: https://cibuildwheel.readthedocs.io/en/stable/faq/#emulation

OK, so I added the aarch64 wheel building job, and it seems to produce a wheel with the correct extension (and also takes much longer). I have no way of verifying that it works, though.

**Comment by lithomas1** [comment]
Also, we should probably figure out how this fits into the release workflow.
Currently, the built wheels are uploaded as artifacts to the GHA workflow.
 
In the other wheel building workflow, wheels are uploaded via a Github release. I'm not too sure how that works(does it make a draft release or actual release?). The main issue with this is that there is no good way to abort the workflow. If a wheel build fails, the other wheels still get uploaded.

For this Action, I think it would be better to write a script that downloads the wheels locally and then manually upload wheels to GH/pypi.

How does this sound? @scoder

File: .github/workflows/wheels.yml (hunk 8 above)
Comment (scoder) [reply .github/workflows/wheels.yml:55]: We can just use the existing workflow then, that's fine.

File: .github/workflows/wheels.yml [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    steps:
+      - name: Checkout Cython
+        uses: actions/checkout@v3
+        
+      - name: Set up QEMU
+        if: matrix.buildplat[1] == 'manylinux_aarch64'
```
Comment (scoder) [inline .github/workflows/wheels.yml:61]:
```suggestion
        if: contains(matrix.buildplat[1], '_aarch64')
```

File: .github/workflows/wheels.yml [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      # Ensure that a wheel builder finishes even if another fails
+      fail-fast: false
+      matrix:
+        # Github Actions doesn't support pairing matrix values together, let's improvise
+        # https://github.com/github/feedback/discussions/7835#discussioncomment-1769026
+        buildplat:
+        - [ubuntu-20.04, manylinux_x86_64]
+        - [ubuntu-20.04, manylinux_aarch64]
+        - [ubuntu-20.04, manylinux_i686]
+        - [ubuntu-20.04, musllinux_x86_64]
```
Comment (scoder) [inline .github/workflows/wheels.yml:51]:
```suggestion
        - [ubuntu-20.04, musllinux_x86_64]
        - [ubuntu-20.04, musllinux_aarch64]
```

File: .github/workflows/wheels.yml [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        - [macos-10.15, macosx_*]
+        - [windows-2019, win_amd64]
+        - [windows-2019, win32]
+        python: ["cp36", "cp37", "cp38", "cp39", "cp310"] # Note: Wheels not needed for PyPy
```
Comment (scoder) [inline .github/workflows/wheels.yml:55]:
```suggestion
        python: ["cp36", "cp37", "cp38", "cp39", "cp310", "cp311"]  # Note: Wheels not needed for PyPy
```

**Comment by scoder** [comment]
> In the other wheel building workflow, wheels are uploaded via a Github release.

I currently create releases manually, which then triggers a fresh wheel build and uploads the wheels to GH-Releases. I then download them from there and send them to PyPI via a script. So there's plenty of time to see if the build succeeds and only upload the wheels if it does.

**Comment by lithomas1** [comment]
This is ready for merge now. I'm not familiar with Github releases or uploading to PyPI, so I'll leave everything past uploading artifacts to you.

**Comment by scoder** [comment]
Thanks!

**Pull request merged**